prober/notification.go

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package prober
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"expvar"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+var (
+	alertGenerated   = expvar.NewInt("alert_generated")
+	alertFailed      = expvar.NewInt("alert_failed")
+	warningGenerated = expvar.NewInt("warning_generated")
+	warningFailed    = expvar.NewInt("warning_failed")
+)
+
+// sendAlert sends an alert to the incident response system, to
+// page a human responder immediately.
+// summary should be short and state the nature of the emergency.
+// details can be longer, up to 29 KBytes.
+func sendAlert(summary, details string) error {
+	type squadcastAlert struct {
+		Message     string            `json:"message"`
+		Description string            `json:"description"`
+		Tags        map[string]string `json:"tags,omitempty"`
+		Status      string            `json:"status"`
+		EventId     string            `json:"event_id"`
+	}
+
+	tags := map[string]string{"severity": "critical"}
+	sqa := squadcastAlert{
+		Message:     summary,
+		Description: details,
+		Tags:        tags,
+		Status:      "trigger",
+		EventId:     uuid.New().String(),
+	}
+
+	sqBody, err := json.Marshal(sqa)
+	if err != nil {
+		alertFailed.Add(1)
+		return fmt.Errorf("encoding alert payload: %w", err)
+	}
+
+	webhookUrl := os.Getenv("SQUADCAST_WEBHOOK")
+	if webhookUrl == "" {
+		warningFailed.Add(1)
+		return errors.New("No SQUADCAST_WEBHOOK configured")
+	}
+
+	req, err := http.NewRequest(http.MethodPost, webhookUrl, bytes.NewBuffer(sqBody))
+	if err != nil {
+		alertFailed.Add(1)
+		return err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{Timeout: 10 * time.Second}
+	resp, err := client.Do(req)
+	if err != nil {
+		alertFailed.Add(1)
+		return err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		alertFailed.Add(1)
+		return errors.New(resp.Status)
+	}
+
+	body, _ := io.ReadAll(resp.Body)
+	if string(body) != "ok" {
+		alertFailed.Add(1)
+		return errors.New("Non-ok response returned from Squadcast")
+	}
+
+	alertGenerated.Add(1)
+	return nil
+}
+
+// sendWarning will post a message to Slack.
+// details should be a description of the issue.
+func sendWarning(details string) error {
+	type slackRequestBody struct {
+		Text string `json:"text"`
+	}
+
+	webhookUrl := os.Getenv("SLACK_WEBHOOK")
+	if webhookUrl == "" {
+		warningFailed.Add(1)
+		return errors.New("No SLACK_WEBHOOK configured")
+	}
+
+	slackBody, err := json.Marshal(slackRequestBody{Text: details})
+	if err != nil {
+		warningFailed.Add(1)
+		return err
+	}
```

Comment:
Nit: I suggest moving the type declaration closer to its use. (I still have a leftover reflex to put declarations at the top from too much C in days of yore 😀)

```suggestion
	webhookUrl := os.Getenv("SLACK_WEBHOOK")
	if webhookUrl == "" {
		warningFailed.Add(1)
		return errors.New("No SLACK_WEBHOOK configured")
	}

	type slackRequestBody struct {
		Text string `json:"text"`
	}

	slackBody, err := json.Marshal(slackRequestBody{Text: details})
	if err != nil {
		warningFailed.Add(1)
		return err
	}
```